### ISSUE TITLE
ios 7 support

### ISSUE THREAD

**Issue by davesegal** [bryankeller/BLKFlexibleHeightBar#4]
Thanks for the great library! Does it support iOS7? I'm getting a consistent crash when I scroll with iOS7 devices

*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Auto Layout still required after executing -layoutSubviews. BLKFlexibleHeightBar's implementation of -layoutSubviews needs to call super.'

**Comment by bryankeller** [comment]
Hi Dave,

I wasn't able to reproduce this issue. Here are the steps I took - tell me what I'm missing and I'll go back and investigate more.

0. Download demo project
1. Change Deployment Target to iOS 7.1
2. Change Base SDK to iOS 7.1
3. Run on iOS 7.1 iPhone 5c simulator

Both the Square Cash demo and Facebook demo worked fine for me. Are you adding the bar in interface builder? If so, it might be getting Auto Layout constraints applied to it, which is probably why you're getting that error.

**Comment by davesegal** [comment]
I have a UIView in IB that has constraints ( in this case a ratio constraint). In code I am adding that view as a subview of the BLKFlexibleHeightBar. Works wonderfully in iOS8 for iOS7 crashes. 

In general do layout constraints conflict with the BLKFlexibleHeightBar?

**Comment by bryankeller** [comment]
Did you add the BLKFlexibleHeightBar as a subview of your table view / scrolling view?

**Comment by davesegal** [comment]
I'm adding the BLKFlexibleHeightBar to the VC's main view like in the example.

**Comment by bryankeller** [comment]
I was able to reproduce the issue by adding a constraints to the bar for one of its subviews. Same crash occurs on iOS 7 but not iOS 8. I'll look into this more soon and hopefully figure out a fix.

Supporting auto layout based layout attributes is a goal for the future too, so this is a good way for me to start looking into that as well.

**Comment by davesegal** [comment]
Thanks!
On a related note, I noticed a strange problem with UISegmentedControl. I am adding a UISegmentedControl as a subview of BLKFlexibleHeightBar. I was able to get BLKFlexibleHeightBar to work on iOS7 when I removed the constraints and used BLKFlexibleHeightBarSubviewLayoutAttributes to define transitional attributes of the UISegmentedControl (in this case only the "center" property). However, on iOS8 the UISegmentControl behaves erratically when setting properties via BLKFlexibleHeightBarSubviewLayoutAttributes. Alternatively UISegmentedControl works perfectly with BLKFlexibleHeightBar and constraints on iOS8 but, again, iOS7 crash.
I'll probably have to find a different solution at this point but let me know if I can help out at all. And thanks for being so responsive.

**Comment by liuxuan30** [comment]
iOS 7 and iOS 8 has different internal implementation for layouting subviews by Apple, which I think is related to this odd issue.
There is a famous error message:
`Auto Layout still required after sending -viewDidLayoutSubviews to the view controller` on iOS 7 device. Your error message looks similar, but slightly different. You can google your error log for a solution.

For `Auto Layout still required after sending -viewDidLayoutSubviews to the view controller` issue:
I solved it like this:
- (void)viewDidLayoutSubviews {
//    CLS_LOG(@"User Setting VC view did layout before super");
    [super viewDidLayoutSubviews];
    // do something
    [self doSomething];

    // for iOS 7 crash. iOS 7 must use this. otherwise you will see Auto Layout still required after sending -viewDidLayoutSubviews to the view controller.
#ifdef __IPHONE_8_0
    if ([[[UIDevice currentDevice] systemVersion] floatValue] < 8.0) {
        [self.view layoutIfNeeded];
    }
#else
    [self.view layoutIfNeeded];

//    CLS_LOG(@"User Setting VC view did layout finished");
}

**Comment by liuxuan30** [comment]
@davesegal, you might want to check your view controller code first and try my solution above to see if you still crash.

**Comment by bryankeller** [comment]
Thanks @liuxuan30 - I think the problem is in the actual bar's code. Quick workaround for now is to change BLKFlexibleHeightBar's -layoutSubviews to this:

```objc
- (void)layoutSubviews
{
    [super layoutSubviews];

    // Update height
    CGRect barFrame = self.frame;
    barFrame.size.height = [self interpolateFromValue:self.maximumBarHeight toValue:self.minimumBarHeight withProgress:self.progress];
    self.frame = barFrame;
    
    if(self.behaviorDefiner && self.behaviorDefiner.isElasticMaximumHeightAtTop)
    {
        self.progress = fmax(self.progress, 0.0);
    }
    
    // Update subviews using the appropriate layout attributes for the current progress
    [self.subviews enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        
        UIView *subview = obj;
        for(int i = 0; i < [subview numberOfLayoutAttributes]; i++)
        {
            CGFloat floorProgressPosition = [subview progressAtIndex:i];
            
            if(i+1 < [subview numberOfLayoutAttributes])
            {
                CGFloat ceilingProgressPosition = [subview progressAtIndex:i+1];
                
                if(self.progress >= floorProgressPosition && self.progress < ceilingProgressPosition)
                {
                    BLKFlexibleHeightBarSubviewLayoutAttributes *floorLayoutAttributes = [subview layoutAttributesAtIndex:i];
                    BLKFlexibleHeightBarSubviewLayoutAttributes *ceilingLayoutAttributes = [subview layoutAttributesAtIndex:i+1];
                    
                    [self applyFloorLayoutAttributes:floorLayoutAttributes ceilingLayoutAttributes:ceilingLayoutAttributes toSubview:subview withFloorProgress:floorProgressPosition ceilingProgress:ceilingProgressPosition];
                }
            }
            else
            {
                if(self.progress >= floorProgressPosition)
                {
                    BLKFlexibleHeightBarSubviewLayoutAttributes *floorLayoutAttributes = [subview layoutAttributesAtIndex:i];
                    
                    [self applyFloorLayoutAttributes:floorLayoutAttributes ceilingLayoutAttributes:floorLayoutAttributes toSubview:subview withFloorProgress:floorProgressPosition ceilingProgress:1.0];
                }
            }
        }
       
        [self layoutIfNeeded];
        
    }];
}
```

Note the 
```objc
[self layoutIfNeeded];
```
at the bottom. I need to investigate this more before I merge a fix into master. I'd like to understand what's actually going on.

**Comment by liuxuan30** [comment]
@bryankeller I strongly recommend you to have a full test on this fix, because `[self layoutIfNeeded]; ` might trigger one more view layout procedure you may not want. Some code will potentially get exeucted one more time and get unwanted results, which I do had. That's why you see I got a macro and a version check, only iOS 7 needs explicit [self layoutIfNeeded]; from what I have tested so far.

**Comment by bryankeller** [comment]
@liuxuan30, agreed. Won't be merging anything into master until I know what's actually going on.

**Comment by karadis** [comment]
It still crashes on ios7 for me. Any updates on the matter?
i'm using a UIView that inherits from BLKFlexibleHeightBar, and has its own NIB.
getting this error: <reason: 'Auto Layout still required after executing -layoutSubviews. WTHomeHeaderView's implementation of -layoutSubviews needs to call super.'>

**Comment by kedio** [comment]
@karadis I get the same error on ios7 and I have not found a work around. Any help appreciated.

**Comment by lekksi** [comment]
+1, having the error on iOS7, iOS8 works well.

**Comment by lekksi** [comment]
Ok, I found a workaround (suboptimal though): I don't define any BLKFlexibleHeightBarSubviewLayoutAttributes on iOS 7 and I'm setting clipsToBounds = YES to hide the top bar subviews when the container view slides into hiding. I hope this helps.
